Eckhart share words and select word components

bieleluk opened the pull request:
this PR:
- renames `new_show_share_words_delizia` to `new_show_share_words_extended` because it will be used by eckhart as well
- creates a full-screen and partial components for showing share words
- creates a full-screen component for selecting a word

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/component/mod.rs:
```diff
@@ -6,18 +6,20 @@ mod header;
 pub use result::{ResultFooter, ResultScreen, ResultStyle};
 #[cfg(feature = "translations")]
 pub use share_words::ShareWords;
+pub use share_words_screen::{ShareWordsScreen, ShareWordsScreenMsg};
```
I'd change it to:
```rust
mod share_words

pub use share_words::{ShareWordsScreen, ShareWordsScreenMsg};
```

I.e.`ShareWords` does not need to be exported via `pub use`, it can live internally within `share_words` module.

bieleluk replied:
https://github.com/trezor/trezor-firmware/pull/4618/commits/92d8ba4c655b1e1f5a055889a1735dc3779e77ff

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+        self.header.render(target);
+        self.content.render(target);
+
+        if let Some(hint) = &self.hint {
```
you can do just `self.hint.render(target)` because `impl Component for Option<T: Component>` exists.

bieleluk replied:
https://github.com/trezor/trezor-firmware/pull/4618/commits/f38328270a125dd3caa004f9ffd872e92f700380

obrusvit reviewed (commented):
I have a few comments for discussion.

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+            self.hint = Some(hint);
+        }
+
+        // if let Some(hint) = &mut self.hint {
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/091ebb6290f2be7c685b37ffc9a5013bd3b9e1c8

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+                    return Some(ShareWordsScreenMsg::Confirmed);
+                }
+                ActionBarMsg::Prev => {
+                    // self.page_idx = (self.page_idx - 1).max(0);
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/091ebb6290f2be7c685b37ffc9a5013bd3b9e1c8

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+        let (header_area, rest) = bounds.split_top(Header::HEADER_HEIGHT);
+        let (rest, action_bar_area) = rest.split_bottom(ActionBar::ACTION_BAR_HEIGHT);
+        let content_area = if let Some(hint) = &mut self.hint {
+            // TODO: hint area based on text
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/091ebb6290f2be7c685b37ffc9a5013bd3b9e1c8

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+        }
+    }
+
+    pub fn with_header(mut self, header: Header) -> Self {
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/5a0abd1fc89be4c258f3cc962dafb15642c7324e

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words_screen.rs:
```diff
@@ -0,0 +1,206 @@
+                    return None;
+                }
+                ActionBarMsg::Next => {
+                    self.content.change_page(self.content.pager().next());
```
This was a bit tricky because it should work for the swipe as well and passing the `event` would result in repeating the code. I solved it it by passing the direction and not calling this function on the `Attach` event  
https://github.com/trezor/trezor-firmware/pull/4618/commits/1fd123bad3a06c7a4929585613f4d49ef8e24721

bieleluk commented on core/mocks/trezortranslate_keys.pyi:
```diff
@@ -761,6 +761,8 @@ class TR:
     setting__adjust: str = "Adjust"
     setting__apply: str = "Apply"
+    share_words__first_word: str = "Write down the first word from the backup."
+    share_words__word_multiple_times: str = "The word appears multiple times in the backup."
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/be40fcf9f00e6c8a8827f68badd3abe59ef9cd31
At the end, the translations should be re-generated to completely remove old translations names

obrusvit commented:
I added two comments to the code, not sure how to access them..
One is here: 
https://github.com/trezor/trezor-firmware/commit/be40fcf9f00e6c8a8827f68badd3abe59ef9cd31#r152505790

bieleluk commented:
The redundant entries `MP_QSTR_share_words__first_word` and `MP_QSTR_share_words__word_multiple_times` are now removed from translations and generated files

bieleluk commented:
- Rebased on top of the `ui-eckhart`
- Implemented _Select word_ component using vertical menu
- added more vertical menu features
- updated button styles and instets according to figma

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/component/vertical_menu.rs:
```diff
@@ -29,19 +29,19 @@ pub struct VerticalMenu {
 impl VerticalMenu {
-    const SIDE_INSET: i16 = 24;
-    const BUTTON_PADDING: i16 = 28;
+    const SIDE_INSET: i16 = 12;
```
Nit: not a fan of calling it `_INSET` if it's not `Inset`.

bieleluk replied:
https://github.com/trezor/trezor-firmware/pull/4618/commits/e2fbcc435ab33ae595d701adb004eae7437fbb33

obrusvit reviewed (commented):
A few comments.

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/component/button.rs:
```diff
@@ -36,7 +36,7 @@ pub struct Button {
 }
 
 impl Button {
-    pub const BASELINE_OFFSET: Offset = Offset::new(2, 6);
```
Based on the offline discussion:
Please remove the constant comletely. Create a `content_offset` field in the `struct Button` instead. Initiate it in the `with_content_offset` builder.

bieleluk replied:
[content_offset](https://github.com/trezor/trezor-firmware/pull/4618/commits/07fabde33993134b9ffdbb5dace3d0c65c115b3b)
https://github.com/trezor/trezor-firmware/pull/4618/commits/99fb21cc0eb998b1b715e4bc0e5f995e211892e2

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words.rs:
```diff
@@ -0,0 +1,342 @@
+pub struct ShareWordsScreen<'a> {
+    header: Header,
+    content: ShareWords<'a>,
+    hint: Option<Hint<'static>>,
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/5030fd7acc6ad92ff9d723a872b331953e0706c5

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/ui_firmware.rs:
```diff
@@ -553,17 +560,20 @@ impl FirmwareUI for UIEckhart {
+        words: heapless::Vec<TString<'static>, 33>,
         _subtitle: Option<TString<'static>>,
         _instructions: Obj,
         _text_footer: Option<TString<'static>>,
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/d0a7c7870faf51daaa8532399779ce365e9d28bb
This is a bit more complicated
- the subtitle is currently not used because there is not a full flow in the FIgma yet
- the instruction is utilized but is only one, because other are dynamically shown in the hint
- text_footer IS `None` in `reset.py` so there is no use for it right now

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/flow/show_share_words.rs:
```diff
@@ -0,0 +1,111 @@
+        match (self, msg) {
+            (Self::Instruction, FlowMsg::Cancelled) => self.return_msg(FlowMsg::Cancelled),
+            (Self::Instruction, FlowMsg::Confirmed) => Self::ShareWords.goto(),
+            (Self::ShareWords, FlowMsg::Cancelled) => self.return_msg(FlowMsg::Cancelled),
```
```rust
            (Self::ShareWords, FlowMsg::Cancelled) => Self::Instructions.goto(),
```

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/component/share_words.rs:
```diff
@@ -0,0 +1,331 @@
+        action_bar.update(content.pager());
+
+        let header = Header::new(TR::reset__recovery_wallet_backup_title.into())
+            .with_right_button(Button::with_icon(theme::ICON_MENU), HeaderMsg::Cancelled);
```
we don't need Menu button here in the `ShareWordsScreen`

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/flow/show_share_words.rs:
```diff
@@ -0,0 +1,111 @@
+        OpTextLayout::new(theme::TEXT_NORMAL).text(text_confirm, fonts::FONT_SATOSHI_REGULAR_38);
+
+    let confirm = TextScreen::new(FormattedText::new(op_confirm))
+        .with_header(Header::new(TR::reset__recovery_wallet_backup_title.into()).with_menu_button())
```
We don't need `menu_button` here.

bieleluk replied:
https://github.com/trezor/trezor-firmware/pull/4618/commits/0baec3112e9b5a727031624f89fbe29b80ff7d61

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/flow/show_share_words.rs:
```diff
@@ -0,0 +1,111 @@
+    let share_words = ShareWordsScreen::new(words).map(|msg| match msg {
+        ShareWordsScreenMsg::Cancelled => Some(FlowMsg::Cancelled),
+        ShareWordsScreenMsg::Confirmed => Some(FlowMsg::Confirmed),
+        ShareWordsScreenMsg::Menu => todo!(),
```
remove this line and others where you have `xyz::Menu => todo!()`

obrusvit replied:
Hmm, I think you can do 
`_ => None`
No need to handle the Menu case.

obrusvit commented on core/embed/rust/src/ui/layout_eckhart/flow/show_share_words.rs:
```diff
@@ -0,0 +1,111 @@
+        .with_header(Header::new(TR::reset__recovery_wallet_backup_title.into()).with_menu_button())
+        .with_action_bar(ActionBar::new_double(
+            Button::with_icon(theme::ICON_CHEVRON_LEFT),
+            Button::with_text(TR::buttons__hold_to_confirm.into())
```
this button can be `.styled(theme::button_confirm())`

bieleluk replied:
https://github.com/trezor/trezor-firmware/pull/4618/commits/1788a16b3c70747767eff61f79d1e2614edbaa48

obrusvit reviewed (commented):
A few more comments.

obrusvit reviewed (approved):
Please finish these nits and then merge.

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/select_word_screen.rs:
```diff
@@ -0,0 +1,107 @@
+    ui::{
+        component::{Component, Event, EventCtx, Label},
+        geometry::{Alignment, Insets, Rect},
+        layout_eckhart::{
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/b603455004b2d066cc5d258b37aa8951840c3645

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/component/share_words.rs:
```diff
@@ -0,0 +1,328 @@
+        },
+        flow::Swipable,
+        geometry::{Alignment, Direction, Offset, Rect},
+        layout_eckhart::{
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/985594154b93e9c870bfcaec601fbc9f159b67fa

bieleluk commented on core/embed/rust/src/ui/layout_eckhart/flow/show_share_words.rs:
```diff
@@ -0,0 +1,111 @@
+            FlowController, FlowMsg, SwipeFlow,
+        },
+        geometry::{Direction, LinearPlacement},
+        layout_eckhart::{
```
https://github.com/trezor/trezor-firmware/pull/4618/commits/0f2bdc182fe8419ab178891c8877086c9429f815
